ipq807x: add support for TP-Link EAP660 HD v1

ElectricGeorge opened the pull request:
Specifications:
* SoC: Qualcomm IPQ8072A (64-bit Quad-core Arm Cortex-A53 @ 1651MHz)
* Memory: 2x ESMT M15T4G16256A-DEBG2G (1 GiB DDR3-1866 13-13-13)
* Serial Port: 3v3 TTL 115200n8
* Wi-Fi: QCN5054 (4x4 5 GHz 802.11ax)
* Wi-Fi: QCN5024 (4x4 2.4 GHz 802.11b/g/n/ax)
* Ethernet: QCA8081 (10/100/1000/2.5GBASE-T)
* Flash: Winbond W29N01HZSINF (128 MiB)
* LEDs: 1x Blue Status (GPIO 42 Active High)
* Buttons: 1x Reset (GPIO 50 Active Low)

Installation Instructions (Serial+TFTP):
1. Solder 4 pin header to JP1 and bridge pads of R58 and R62.
2. Connect 3V3 TTL port to TX, RX, and GND, which are positions 1, 2, and 3 respectively. Be sure to crossover TX and RX.
3. Copy RAM firmware image openwrt-qualcommax-ipq807x-tplink_eap660hd-v1-initramfs-uImage.itb to TFTP server root, available at 192.168.10.1.
4. Connect PoE ethernet cable to the RJ45 port and hold Ctrl+B in the serial console (115200 baud) until autoboot is halted.
5. Run the following commands in the U-boot prompt: 
    ```
    tftpboot 0x44000000 openwrt-qualcommax-ipq807x-tplink_eap660hd-v1-initramfs-uImage.itb
    bootm
    ```
    You may need to type Ctrl+C and Enter before running these commands to clear invisible characters from the buffer.
6. Run the following command in a terminal to copy the sysupgrade image to be installed:
    ```
    scp openwrt-qualcommax-ipq807x-tplink_eap660hd-v1-squashfs-sysupgrade.bin root@192.168.1.1:/tmp/
    ```
7. Activate the OpenWrt serial console and run the following commands:
    ```
    cd /tmp
    sysupgrade -n openwrt-qualcommax-ipq807x-tplink_eap660hd-v1-squashfs-sysupgrade.bin
    ```
8. The AP will reboot and OpenWrt will be successfully installed.

ElectricGeorge commented:
The new TP-Link factory image format is currently a work in progress. Some scripts were copied from EAP610-Outdoor. See #14922.

robimarko commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-eap660hd-v1.dts:
```diff
@@ -0,0 +1,131 @@
+	phy-handle = <&qca8081_28>;
+	qcom,id = <5>;
+	qcom,phy-mdio-addr = <28>;
+};
```
This node is wrong, you dont want `dp1`, from the looks you want `dp5` and then you can drop `qcom,id`.

Drop `qcom,phy-mdio-addr = <28>` as that is not used.

robimarko commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-eap660hd-v1.dts:
```diff
@@ -0,0 +1,131 @@
+	switch_mac_mode2 = <MAC_MODE_SGMII_CHANNEL0>;
+
+	qcom,port_phyinfo {
+		port@4 {
```
port@5

robimarko commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-eap660hd-v1.dts:
```diff
@@ -0,0 +1,131 @@
+&switch {
+	status = "okay";
+	switch_lan_bmp = <ESS_PORT5>;
+	switch_mac_mode = <MAC_MODE_PSGMII>;
```
You are not using this UNIPHY at all, just remove the entry, as its disabled by default.

robimarko commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-eap660hd-v1.dts:
```diff
@@ -0,0 +1,131 @@
+	status = "okay";
+	switch_lan_bmp = <ESS_PORT5>;
+	switch_mac_mode = <MAC_MODE_PSGMII>;
+	switch_mac_mode1 = <MAC_MODE_SGMII_CHANNEL0>;
```
You want `MAC_MODE_SGMII_PLUS` here

robimarko commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-eap660hd-v1.dts:
```diff
@@ -0,0 +1,131 @@
+	switch_lan_bmp = <ESS_PORT5>;
+	switch_mac_mode = <MAC_MODE_PSGMII>;
+	switch_mac_mode1 = <MAC_MODE_SGMII_CHANNEL0>;
+	switch_mac_mode2 = <MAC_MODE_SGMII_CHANNEL0>;
```
Again, this UNIPHY is unused, just drop it.

robimarko commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-eap660hd-v1.dts:
```diff
@@ -0,0 +1,131 @@
+		nand-ecc-strength = <4>;
+		nand-ecc-step-size = <512>;
+		nand-bus-width = <8>;
+	};
```
I guess SMEM is populating this?

If so, please add the smem compatible so that is clear

robimarko commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-eap660hd-v1.dts:
```diff
@@ -0,0 +1,131 @@
+	reset-gpios = <&tlmm 25 GPIO_ACTIVE_LOW>;
+	
+	qca8081_28: ethernet-phy@28 {
+		reg = <0x1c>;
```
Use the decimal value here, and add the PHY compatible.

robimarko commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-eap660hd-v1.dts:
```diff
@@ -0,0 +1,131 @@
+		compatible = "gpio-leds";
+
+		led_status_blue: status_blue {
+			label = "blue:status";
```
No, please switch to color and function instead and drop label.

ElectricGeorge replied:
I'm not sure what you mean. Isn't blue:status color and function? Are you referring to the DT label `led_status_blue` or the `label` property?

robimarko replied:
No, the label property is deprecated.

Now we are using function and color DT properties, look at this board for example:
https://github.com/openwrt/openwrt/blob/main/target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8074-sxk80.dtsi

There is plenty of others also using color/function combo.

robimarko commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-eap660hd-v1.dts:
```diff
@@ -0,0 +1,131 @@
+
+/ {
+	model = "TP-Link EAP660 HD v1";
+	compatible = "tplink,eap660hd-v1", "qcom,ipq807x-hk07", "qcom,ipq807x";
```
Do you really need the `"qcom,ipq807x-hk07"` compatible?

Also, please change `"qcom,ipq807x"` to `"qcom,ipq8074"`

robimarko commented on target/linux/qualcommax/ipq807x/base-files/etc/board.d/02_network:
```diff
@@ -9,6 +9,7 @@
 ipq807x_setup_interfaces()
 {
 	local board="$1"
+	local macaddr
```
No need for this as MAC setting should happen in `ipq807x_setup_macs`

robimarko commented on target/linux/qualcommax/ipq807x/base-files/etc/board.d/02_network:
```diff
@@ -54,6 +55,11 @@ ipq807x_setup_interfaces()
 		;;
+	tplink,eap660hd-v1)
+		macaddr=$(get_mac_binary /tmp/factory_data/default-mac 0)
+		ucidef_set_interface_macaddr "lan" "$macaddr"
```
MAC setting should happen in `ipq807x_setup_macs`

robimarko commented on target/linux/qualcommax/ipq807x/base-files/etc/hotplug.d/firmware/11-ath11k-caldata:
```diff
@@ -72,6 +72,9 @@ case "$FIRMWARE" in
 		ath11k_set_macflag
 		;;
+	tplink,eap660hd-v1)
+		cp /tmp/factory_data/radio "/lib/firmware/$FIRMWARE"
```
`caldata_from_file` should work here instead

robimarko commented on target/linux/qualcommax/ipq807x/base-files/lib/preinit/09_mount_factory_data:
```diff
@@ -0,0 +1,19 @@
+		mtd_path=$(find_mtd_chardev "factory_data")
+		ubiattach --dev-path="$mtd_path" --devn=1
+		mkdir /tmp/factory_data
+		mount -o ro,noatime -t ubifs ubi1:ubi_factory_data /tmp/factory_data
```
Can you share the file structure that this FS contains?

ElectricGeorge replied:
```
-rw-rw-r--    1 500      500              4 Nov 20  2020 add-hardver
-r-sr-Sr-x    1 root     root             6 Jan  1  2020 default-mac
--wS--sr-x    1 root     root            40 Jan  1  2020 device-id
-rwxr-xr-x    1 root     root            74 Jan  1 00:00 device-info
--ws--x--t    1 root     root        131072 Jan  1  2020 radio
-r-sr-Sr-x    1 root     root             8 Jan  1  2020 region
```

robimarko commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-eap660hd-v1.dts:
```diff
@@ -0,0 +1,133 @@
+	reset-gpios = <&tlmm 25 GPIO_ACTIVE_LOW>;
+	
+	qca8081_28: ethernet-phy@28 {
+		compatible = "ethernet-phy-id004d.d101";
```
Regular C22 compatible is fine since you dont have a reset-gpios attached

aiamadeus commented on package/firmware/ipq-wifi/Makefile:
```diff
@@ -45,6 +45,7 @@ ALLWIFIBOARDS:= \
 	netgear_wax218 \
 	netgear_wax620 \
 	netgear_wax630 \
+	tplink_eap660hd-v1 \
```
Keep alphabetical order.

aiamadeus commented on package/firmware/ipq-wifi/Makefile:
```diff
@@ -167,6 +168,7 @@ $(eval $(call generate-ipq-wifi-package,netgear_wax214,Netgear WAX214))
 $(eval $(call generate-ipq-wifi-package,netgear_wax218,Netgear WAX218))
 $(eval $(call generate-ipq-wifi-package,netgear_wax620,Netgear WAX620))
 $(eval $(call generate-ipq-wifi-package,netgear_wax630,Netgear WAX630))
+$(eval $(call generate-ipq-wifi-package,tplink_eap660hd-v1,TP-Link EAP660 HD v1))
```
Some here.

aiamadeus commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-eap660hd-v1.dts:
```diff
@@ -0,0 +1,134 @@
+			gpios = <&tlmm 42 GPIO_ACTIVE_HIGH>;
+			color = <LED_COLOR_ID_BLUE>;
+		};
+
```
Extra blank lines.

aiamadeus commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-eap660hd-v1.dts:
```diff
@@ -0,0 +1,134 @@
+	reset-gpios = <&tlmm 25 GPIO_ACTIVE_LOW>;
+	
+	qca8081_28: ethernet-phy@28 {
+		compatible = "ethernet-phy-ieee802.3-c22";
```
This is a bit weird.

ElectricGeorge replied:
How so? I'm not completely familiar with this project, so forgive me

robimarko replied:
Its not weird, QCA8081 is C22 compatible.

aiamadeus commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-eap660hd-v1.dts:
```diff
@@ -0,0 +1,134 @@
+	status = "okay";
+	qcom,ath11k-calibration-variant = "TP-Link-EAP660-HD-v1";
+};
+
```
Extra blank lines.

aiamadeus commented on target/linux/qualcommax/image/ipq807x.mk:
```diff
@@ -259,6 +259,20 @@ define Device/prpl_haze
 endef
 TARGET_DEVICES += prpl_haze
 
+define Device/tplink_eap660hd-v1
```
Keep alphabetical order.

aiamadeus commented on target/linux/qualcommax/ipq807x/base-files/etc/board.d/02_network:
```diff
@@ -54,6 +54,9 @@ ipq807x_setup_interfaces()
 	netgear,wax630)
 		ucidef_set_interface_lan "lan1 lan2" "dhcp"
 		;;
+	tplink,eap660hd-v1)
```
Some here.

aiamadeus commented on target/linux/qualcommax/ipq807x/base-files/etc/hotplug.d/firmware/11-ath11k-caldata:
```diff
@@ -72,6 +72,9 @@ case "$FIRMWARE" in
 		ath11k_patch_mac $(macaddr_add $label_mac 33) 2
 		ath11k_set_macflag
 		;;
+	tplink,eap660hd-v1)
```
Some here.

aiamadeus commented on target/linux/qualcommax/ipq807x/base-files/lib/upgrade/platform.sh:
```diff
@@ -56,6 +129,9 @@ platform_do_upgrade() {
 	netgear,wax630)
 		nand_do_upgrade "$1"
 		;;
+	tplink,eap660hd-v1)
```
Some here.

aiamadeus reviewed (commented):
> SoC: Qualcomm IPQ8072A (64-bit Quad-core Arm Cortex-A53 @ 1651MHz)

This is a bit strange, isn't the IPQ8072 SoC 2.2GHz?

ElectricGeorge committed:
ipq807x: add support for TP-Link EAP660 HD v1

Specifications:
* SoC: Qualcomm IPQ8072A (64-bit Quad-core Arm Cortex-A53 @ 2200MHz)
* Memory: 2x ESMT M15T4G16256A-DEBG2G (1 GiB DDR3-1866 13-13-13)
* Serial Port: 3v3 TTL 115200n8
* Wi-Fi: QCN5054 (4x4 5 GHz 802.11ax)
* Wi-Fi: QCN5024 (4x4 2.4 GHz 802.11b/g/n/ax)
* Ethernet: QCA8081 (10/100/1000/2.5GBASE-T)
* Flash: Winbond W29N01HZSINF (128 MiB)
* LEDs: 1x Blue Status (GPIO 42 Active High)
* Buttons: 1x Reset (GPIO 50 Active Low)

Installation Instructions (Serial+TFTP):
1. Solder 4 pin header to JP1 and bridge pads of R58 and R62.
2. Connect 3V3 TTL port to TX, RX, and GND, which are positions 1, 2,
   and 3 respectively. Be sure to crossover TX and RX.
3. Copy RAM firmware image
   openwrt-qualcommax-ipq807x-tplink_eap660hd-v1-initramfs-uImage.itb
   to TFTP server root, available at 192.168.10.1.
4. Connect PoE ethernet cable to the RJ45 port and hold Ctrl+B in the
   serial console (115200 baud) until autoboot is halted.
5. Run the following commands in the U-boot prompt:
   # tftpboot 0x44000000 openwrt-qualcommax-ipq807x-tplink_eap660hd-v1-initramfs-uImage.itb
   # bootm
   You may need to type Ctrl+C and Enter before running these commands
   to clear invisible characters from the buffer.
6. Run the following command in a terminal to copy the sysupgrade image
   to be installed (check IP address):
   $ scp openwrt-qualcommax-ipq807x-tplink_eap660hd-v1-squashfs-sysupgrade.bin root@192.168.1.1:/tmp/
7. Activate the OpenWrt serial console and run the following commands:
   # cd /tmp
   # sysupgrade -n openwrt-qualcommax-ipq807x-tplink_eap660hd-v1-squashfs-sysupgrade.bin
8. The AP will reboot and OpenWrt will be successfully installed.

Signed-off-by: George Witt <george.witt@nltsproject.org>
Link: https://github.com/openwrt/openwrt/pull/15832
Signed-off-by: Robert Marko <robimarko@gmail.com>

ElectricGeorge commented:
> > SoC: Qualcomm IPQ8072A (64-bit Quad-core Arm Cortex-A53 @ 1651MHz)
> 
> This is a bit strange, isn't the IPQ8072 SoC 2.2GHz?

You are correct. The 1651 MHz was from kernel logs of the stock firmware. I have updated the commit message

robimarko commented on target/linux/qualcommax/ipq807x/base-files/etc/hotplug.d/ieee80211/11_fix_wifi_mac:
```diff
@@ -23,4 +23,8 @@ case "$board" in
+	tplink,eap660hd-v1)
+		base_mac=$(get_mac_binary /tmp/factory_data/default-mac 0)
+		macaddr_add $base_mac $(expr "$PHYNBR" + 1) > /sys${DEVPATH}/macaddress
+		;;
```
One option is to use the recent ath11k MAC patching support to patch the MAC to caldata instead of setting it later.

ElectricGeorge replied:
After further review of the stock image, the ethernet NIC and the 2.4GHz radio share the same MAC. I made the above change with that in mind. Is that okay, or should there be a different address for each NIC?

robimarko replied:
MAC-s should reflect those in the stock FW, so if stock FW reuses the same then fine.

robimarko commented:
Any chance you have time to respin this?

ElectricGeorge commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-eap660hd-v1.dts:
```diff
@@ -0,0 +1,132 @@
+	compatible = "tplink,eap660hd-v1", "qcom,ipq8074";
+
+	aliases {
+		ethernet0 = &dp5;
```
Looks like the MAC was invalid in U-boot

professor-jonny reviewed (approved):
I did notice the BDF file submitted is named IPQ8074 and the target is a IPQ8072, it actually looks like all others are named the same, more of a question upstream than an issue with this pull request.

Other than that, looks good.

by chance do you have the smem partition layout or bootlog to look at possibly on the forum or wiki?

https://github.com/openwrt/firmware_qca-wireless/blob/main/board-tplink_eap660hd-v1.ipq8074

robimarko commented:
Thanks! Rebased on top of main and merged!
